prow/pjutil/abort.go

```diff
@@ -36,35 +37,79 @@ type prowClient interface {
 // message.
 type ProwJobResourcesCleanup func(pj prowapi.ProwJob) error
 
+func refsForRefs(ref prowapi.Refs) refs {
+	var pulls []int
+	for _, pull := range ref.Pulls {
+		pulls = append(pulls, pull.Number)
+	}
+	return refs{
+		org:   ref.Org,
+		repo:  ref.Repo,
+		base:  ref.BaseRef,
+		pulls: pulls,
+	}
+}
+
 // jobIndentifier keeps the information required to uniquely identify a prow job
 type jobIndentifier struct {
-	job          string
-	organization string
-	repository   string
-	pullRequest  int
+	job  string
+	refs []refs
+}
+
+type refs struct {
+	org, repo, base string
+	pulls           []int
 }
 
 // String returns the string representation of a prow job identifier
-func (i *jobIndentifier) String() string {
-	return fmt.Sprintf("%s %s/%s#%d", i.job, i.organization, i.repository, i.pullRequest)
+func (i jobIndentifier) String() string {
+	description := fmt.Sprintf("%s ", i.job)
+	for _, ref := range i.refs {
+		description += fmt.Sprintf("%s/%s@%s %v", ref.org, ref.repo, ref.base, ref.pulls)
+	}
+	return description
 }
 
-// TerminateOlderPresubmitJobs aborts all presubmit jobs from the given list that have a newer version. It calls
+// TerminateOlderJobs aborts all presubmit jobs from the given list that have a newer version. It calls
 // the cleanup callback for each job before updating its status as aborted.
-func TerminateOlderPresubmitJobs(pjc prowClient, log *logrus.Entry, pjs []prowapi.ProwJob,
+func TerminateOlderJobs(pjc prowClient, log *logrus.Entry, pjs []prowapi.ProwJob,
 	cleanup ProwJobResourcesCleanup) error {
-	dupes := map[jobIndentifier]int{}
+	dupes := map[string]int{}
 	for i, pj := range pjs {
-		if pj.Complete() || pj.Spec.Type != prowapi.PresubmitJob {
+		if pj.Complete() || !(pj.Spec.Type == prowapi.PresubmitJob || pj.Spec.Type == prowapi.BatchJob) {
 			continue
 		}
 
-		ji := jobIndentifier{
-			job:          pj.Spec.Job,
-			organization: pj.Spec.Refs.Org,
-			repository:   pj.Spec.Refs.Repo,
-			pullRequest:  pj.Spec.Refs.Pulls[0].Number,
+		// we want to use the jobIdentifier as a key to a map,
+		// so we must make sure that the equality and comparison
+		// operators are doing what we want in the map creation.
+		// Namely, we must sort all of the slices in our identifier
+		// so that equivalent permutations of the refs map to the
+		// same identifier. We do not want commit hashes to matter
+		// here as a test for a newer set of commits but for the
+		// same set of names can abort older versions
+		var refs []refs
+		if pj.Spec.Refs != nil {
+			refs = append(refs, refsForRefs(*pj.Spec.Refs))
+		}
+		for _, ref := range pj.Spec.ExtraRefs {
+			refs = append(refs, refsForRefs(ref))
+		}
+
+		for i := range refs {
+			sort.Ints(refs[i].pulls)
 		}
+		sort.Slice(refs, func(i, j int) bool {
+			if refs[i].org != refs[j].org {
+				return refs[i].org < refs[j].org
+			}
+			return refs[i].repo < refs[j].repo
```

Comment:
If there are multiple refs that reference the same repo their ordering will be undefined. We should probably also sort based on the prs.
Defining an injective `String` method or function for refs would make it easy to sort (`return refs[i].String() < refs[j].String()`) and could be reused in the `jobIdentifier.String` method.

Reply:
ACK